dask/task_spec.py

```diff
@@ -0,0 +1,905 @@
+        return self
+
+    def __getstate__(self):
+        self.pack()
+        return {
+            "key": self.key,
+            "packed_func": self.packed_func,
+            "dependencies": self.dependencies,
+            "kwargs": self.kwargs,
+            "args": self.args,
+            "_is_coro": self._is_coro,
+            "_token": self._token,
+        }
+
+    def __setstate__(self, state):
+        self.key = state["key"]
+        self.packed_func = state["packed_func"]
+        self.dependencies = state["dependencies"]
+        self.kwargs = state["kwargs"]
+        self.args = state["args"]
+        self._is_coro = state["_is_coro"]
+        self._token = state["_token"]
+        self.func = None
+
+    def __eq__(self, value: object) -> bool:
+        if not isinstance(value, Task):
+            return False
+        if self.key != value.key:
+            return False
+        if self.packed_func != value.packed_func:
+            return False
+        if self.func != value.func:
+            return False
+        if self.dependencies != value.dependencies:
+            return False
+        if self.kwargs != value.kwargs:
+            return False
+        if self.args != value.args:
+            return False
+        return True
+
+    @property
+    def is_coro(self):
+        if self._is_coro is None:
+            self.unpack()
+            # Note: Can't use cached_property on objects without __dict__
+            try:
+                from distributed.utils import iscoroutinefunction
+
+                self._is_coro = iscoroutinefunction(self.func)
+            except Exception:
+                self._is_coro = False
+        return self._is_coro
+
+
+class DictOfTasks(BaseTask):
+    tasks: dict
+    __slots__ = tuple(__annotations__)
+
+    def __init__(self, key: Any, nested_tasks: dict):
+        self.key = key
+        self.dependencies = set()
+        self.tasks = {}
+        for k, task in nested_tasks.items():
+            self.tasks[k] = task = parse_input(task)
+            if isinstance(task, BaseTask):
+                self.dependencies.update(task.dependencies)
+
+    def inline(self, dsk) -> DictOfTasks:
+        new_tasks = {k: task.inline(dsk) for k, task in self.tasks.items()}
+        return DictOfTasks(self.key, new_tasks)
+
+    def __call__(self, values=()) -> dict:
+        self._verify_values(values)
+        return {
+            k: task(values) if isinstance(task, BaseTask) else task
+            for k, task in self.tasks.items()
+        }
+
+    def propagate_literal(self) -> BaseTask:
+        """Return a literal if all tasks are literals"""
+        if any(
+            isinstance(t, BaseTask) and not isinstance(t, LiteralTask)
+            for t in self.tasks.values()
+        ):
+            return self
+        return LiteralTask(self.key, self())
+
+    def __sizeof__(self):
+        return super().__sizeof__() + sizeof(self.tasks)
+
+
+class SequenceOfTasks(BaseTask):
```

Comment:
Note that I opted to keep this after all. It is not something end users (or even developers) really have to touch but it makes internals of the Task class much smoother. I'm not married to it, though, and could see it being removed down the line again. It would just require a lot more iterating over things, checking for types, etc.

Reply:
I'm not necessarily against these classes, I mostly dislike their name which suggests less magic than they actually perform. This might be rather something like a `Sequence{Consolidation|Conflation|Combination|Comprehension|Reduction}Task`